src/parser.mly

```diff
@@ -0,0 +1,142 @@
+/* Ocamlyacc parser for SIFT */
+%{ 
+  open Ast 
+%}
+
+/* Arithmetic Operators */
+%token ADD SUBTRACT MULTIPLY DIVIDE MOD
+/* Relational Operators */
+%token EQ LEQ GT GEQ NEQ LT
+/* Logical Operators */
+%token AND OR NOT XOR
+/* Paranthesis and Brackets */
+%token LPAREN RPAREN LBRACE RBRACE LARROW RARROW LSQBRACE RSQBRACE
+/* Delimiters */
+%token COMMA SEMICOLON COLON DOT
+/* Assignment */
+%token ASSIGN 
+/* Features */
+%token PIPE
+/* Keywords */
+%token IF ELIF ELSE SWITCH CASE FOR WHILE CONTINUE BREAK DEFAULT LAMBDA DEFINE EXIT RETURN NULL TRUE FALSE
+%token IMPORT LIST FROM PURE 
+/* Primitive Data Types */
+%token INT FLOAT CHAR SYMBOL STRING BOOL
+/* Non-primitive Data Types */
+%token ARRAY LIST TUPLE SET DICT
+%token <int> INT_LITERAL
+%token <float> FLOAT_LITERAL
+%token <bool> BLIT
+%token <string> ID
+%token EOF
+
+
+%left SEMICOLON
+%nonassoc RSQBRACE RBRACE RARROW RPAREN
+%nonassoc LSQBRACE LBRACE LARROW LPAREN
+%nonassoc COLON
+%nonassoc COMMA
+%nonassoc NOELSE
+%left RETURN
+%right ASSIGN
+%left OR AND
+%left EQ NEQ
+%left GT GEQ LEQ LT
+%left ADD SUBTRACT
+%left MULTIPLY DIVIDE MOD
+%right NOT
+
+%start program
+%type <Ast.program> program
+
+%right ASSIGN
+%left OR
+%left AND
+%left EQ NEQ
+%left LT
+%left PLUS MINUS
+
+%%
+
+/* add function declarations*/
+program:
+  decls EOF { $1}
+
+decls:
+   /* nothing */ { ([], [])               }
+ | vdecl SEMI decls { (($1 :: fst $3), snd $3) }
+ | fdecl decls { (fst $2, ($1 :: snd $2)) }
+
+vdecl_list:
+  /*nothing*/ { [] }
+  | vdecl SEMI vdecl_list  {  $1 :: $3 }
+
+/* int x */
+vdecl:
+  typ ID { ($1, $2) }
+
+typ:
+    INT   { Int   }
+  | BOOL  { Bool  }
+
+/* fdecl */
+fdecl:
+  vdecl LPAREN formals_opt RPAREN LBRACE vdecl_list stmt_list RBRACE
+  {
+    {
+      rtyp=fst $1;
+      fname=snd $1;
+      formals=$3;
+      locals=$6;
+      body=$7
+    }
+  }
+
+/* formals_opt */
+formals_opt:
+  /*nothing*/ { [] }
+  | formals_list { $1 }
+
+formals_list:
+  vdecl { [$1] }
+  | vdecl COMMA formals_list { $1::$3 }
+
+stmt_list:
+  /* nothing */ { [] }
+  | stmt stmt_list  { $1::$2 }
+
+stmt:
+    expr SEMI                               { Expr $1      }
+  | LBRACE stmt_list RBRACE                 { Block $2 }
```

Comment:
What is this supposed to be?

Reply:
This is just a copy of nanoparse. Haven't changed it.